### DISCUSSION TITLE
Can Windows Boot Manager Be Bypassed?

### DISCUSSION THREAD

**Discussion by hugofant** [dakanji/RefindPlus#197] [General]
Hi,
i've installed Windows 11 to two different partitions (one for working, one for gaming). I'm booting UEFI with secure boot disabled.
Now i want to use RefindPlus for starting them.....my problem: Refind detects the windows boot manager, there i've to select the windows partition....and then it starts Refind again and when selecting windows again, its start the previous selected version of windows.
Is there a way to bypass windows boot manager, for example booting winload.efi directly from partition?

**Comment by dakanji** [comment]
You should be able to load Windows directly ... if that is what you are asking about.
If you cannot see where/how to do this, please boot the DBG version and share the log it produces.
This will be in the same ESP you ran RefindPlus from.

**Reply by hugofant** [reply]
Thanks for your answer/support:
here are the logs - 2 of them. as said refind autodetects the windows bootmanager (on EFI partition), after selecting the windows partition there it boots again refind, selecting again windows bootmanager start the windows partition selected before.

[24v18k0604.log](https://github.com/user-attachments/files/17797719/24v18k0604.log)
[24v18k0636.log](https://github.com/user-attachments/files/17797721/24v18k0636.log)

**Comment by dakanji** [comment]
Got it. I was a bit loose with my words when I said you can load Windows directly, as no, you currently cannot in the way you meant.
That is, RefindPlus always loads the Windows Boot Manager on the ESP for UEFI Windows and does not try to run files from within Windows.

The log shows RefindPlus is working as expected. I cannot offer anything on why the Windows Boot Manager is not doing what it is meant to.
Perhaps try disconnecting one of the Windows instances each to see if there is any difference in the outcome.

**Comment by hugofant** [comment]
ok...what i don't understand: the first time the windows bootmanager is entered by refind it shows the selection menu. the second time it boots directly into the windows partition selected before.....so is't it possible to call the windows boot manager for example with a parameter to tell it to boot directly a specified partition (as done the second time its entered)?

**Comment by dakanji** [comment]
When you say "refind", do you mean RefindPlus or rEFInd?
Best to spell precisely to eliminate confusion.

In terms of params, yes it is possible to pass params to items being called but RefindPlus does not pass any params to UEFI Windows. 
That is, any difference in behavior is not down to parameters being passed:

![000-Snip1](https://github.com/user-attachments/assets/9aa2c1cf-c215-4ba1-a5de-4e6937df49dc)
![000-Snip2](https://github.com/user-attachments/assets/a45d192e-48ad-434b-8b58-3e828ce6ee1b)

This is unless you meant rEFInd but I don't think that passes any params either

**Comment by dakanji** [comment]
> That is, any difference in behavior is not down to parameters being passed

Well, on thinking about things a bit, while RefindPlus is not passing any params, it does not mean there no params in play as you summised.

What may be happening is that the Boot Manager could be looking for a `Boot####` nvRAM variable, then setting this and returning to whatever started it if absent, and then, if run again, picking this up and using it.

RefindPlus could set this for the first boot if what it is is known. You can try the following to check:
- Load RefindPlus and load uEFI shell
- Run `dh -d -v > dh_d_v_before.txt`
- Exit shell and load UEFI Windows
- When you are returned to Refindplus, load the shell again and run `dh -d -v > dh_d_v_after.txt`
- Exit the shell and continue

Share the files to see what, if anything, changed between the two runs.

NB: You probably need to enter `FS0:` before running those shell commands
The RefindPlus package may contain a UEFI shell file but I forget.
If it does not, you can get one here: https://github.com/tianocore/edk2/tree/UDK2018/ShellBinPkg/UefiShell/X64. 
See here for HowTo: https://github.com/dakanji/RefindPlus/discussions/105#discussioncomment-6157661

**Comment by hugofant** [comment]
Hi again,
to be more precisely - yes i'm talking about "refind PLUS". sorry for not being clear enough.
Have implemented shell and executed the commands suggested by you....there are some differences, but i'm not able to interpret if they are relevent or not.
Files are attached.
[dh_d_v_after.txt](https://github.com/user-attachments/files/17815873/dh_d_v_after.txt)
[dh_d_v_before.txt](https://github.com/user-attachments/files/17815874/dh_d_v_before.txt)

**Comment by dakanji** [comment]
Thanks but sorry, I gave you a command for different output.

Please use these instead:
- dmpstore > nvram_dump_before.txt
- dmpstore > nvram_dump_after.txt

I will still look that the ones from earlier (they show what images are loaded)

PS: Correct spelling is `RefindPlus` for this and for the upstream project, `rEFInd`

**Comment by hugofant** [comment]
hmm, not so good news - both file are identical. So it seems that Windows Boot Manager handles this "internally"

**Comment by dakanji** [comment]
I see. Well, it was just a wild idea to potentially work around what seems to be a broken/misconfigured setup.

The Windows boot manager should not be loading something else on selecting a Windows instance it presents. You need to look into how and why this is so and fix it. Best bet is on a Windows support channel. 

Good luck!

**Reply by hugofant** [reply]
> what seems to be a broken/misconfigured setup

no, there's nothing broken/misconfigured

> The Windows boot manager should not be loading something else on selecting a Windows instance it presents

seems not to be correct.
In the meantime i've found a [site](https://www.toptensoftware.com/blog/booting-multiple-windows-with-refind/) which explains the bootprocess (same as mine)....and also describes a "work araound" for my problem.
I'll try this and will give a feedback

**Comment by dakanji** [comment]
The site's sugested fix makes sense. Technically though, it is fixing a "misconfiguration" from the Windows boot manager's perpective.

Before you make the changes suggested there, try to save copies of the BCD file before and after to see whether this is how the "Params" are being passed (I suspect it is). It might be possible to add something to RefindPlus to read/amend the BCD as needed if so.

**Comment by hugofant** [comment]
mentioned method works well :-))  WindowsBootManager starts windows directly, when there is only one posibility available ...and have hidden the original MicrosoftBootManager.
if you're interested in - BCDs are attached (had to rename them to *.zip as upload without extensions was not possible)
[BCD_both.zip](https://github.com/user-attachments/files/17832600/BCD_both.zip)
[BCD_gaming.zip](https://github.com/user-attachments/files/17832601/BCD_gaming.zip)
[BCD_working.zip](https://github.com/user-attachments/files/17832602/BCD_working.zip)

**Comment by dakanji** [comment]
Thanks. Nice.

What I was actually thinking might be useful was to go through a version of the steps I outlined for the nvram dump but with the orignal BCD instead. The boot manager must be amending this as the BCD is effectively an extension of the nvRAM. Can then try to figure out what is changed.

- Load RefindPlus, unhide the orignal windows entry and load uEFI shell
- Run `cp /Path/To/Orignal/BCD /Path/To/Orignal/BCD_BEFORE`
- Exit shell and start load UEFI Windows with the previously hidden instance
- Select same Windows instance that triggered the previous behaviour 
- When you are returned to Refindplus, load the shell again and run `cp /Path/To/Orignal/BCD /Path/To/Orignal/BCD_AFTER`
- Exit shell and continue

Basically trying to see what gets changed.
Can perhaps then, at some point, look at modifying RefindPlus to make similar changes before trying to load the boot manager.
